proposals/pattern-variables.md

```diff
@@ -0,0 +1,150 @@
+
+	```cs
+	e is not (0, var x) and not (var x, 0)
+	e is not (0, var x) && e is not (var x, 0)
+	```
+   Note: This is the DeMorgan's transformation of the previous case.
+
+For a *logical_not_expression* of the form `!expr_operand`:
+- Pattern variables declared in the *expr_operand* can be redeclared anywhere in the containing expression; permitting:
+
+  ```cs
+  !(e is (0, var x)) && !(e is (var x, 0))
+  ```
+  Note: This is the expression variation of the previous case.
+
+For a *switch_case_label* of the form `case case_pattern when when_expr`:
+- Pattern variables declared in the *case_pattern* or *when_expr* can be redeclared in other case labels; permitting:
+
+	```cs
+	case (var x, 0) when e is int i:
+	case (0, var x) when a is int i:
+	```
+
+	Note: This rule is currently defined for each individual *switch_section* only.
+	
+For a *conditional_expression* of the form `expr_cond ? expr_true : expr_false`:
+- Pattern variables declared in *expr_true* can be redeclared in *expr_false*; permitting:
+
+	```cs
+	b ? x is int i : y is int i
+	```
+
+Redeclaring pattern variables is only permitted for variables of the same type.
+
+> **Open question**: How identical these types should be?
+
+Redeclaring pattern variables in any other case follows the usual scoping rules and is disallowed.
+
+### Definite assignment
+
+Note: This section is unchanged and included for the sake of completeness.
+
+For an *is_pattern_expression* of the form `e is pattern`:
+
+- The definite assignment state of *v* after *is_pattern_expression* is determined by:
+
+    - The state of *v* after *is_pattern_expression* is definitely assigned, if the *pattern* is irrefutable. permitting use after:
+    
+        ```cs
+        _ = x is var x;
+        _ = 1 is int x;
+        _ = (1, 2) is var (x, y) and var z;
+       ```
+
+    - Otherwise, the state of *v* after *is_pattern_expression* is the same as the state of *v* after *pattern*.
+
+For a *var_pattern* of the form `var variable_designation`:
+
+- The definite assignment state of *v* after *var_pattern* is determined by:
+
+    - The state of *v* after *var_pattern* is definitely assigned if *variable_designation* is a *single_variable_designation*.
+    - Otherwise, the state of *v* after *var_pattern* is "definitely assigned when true".
+
+#### General rules for pattern variables in simple patterns 
+
+Note: This section is unchanged and included for the sake of completeness.
+
+The following rules applies to any *primary_pattern* that declares a variable:
+
+- The state of *v* is "definitely assigned when true" after *primary_pattern*; permitting use after:
+
+    ```cs
+    if (o is int x)
+    ```
+
+#### Definite assignment rules for pattern variables in pattern combinators
+
+Note: This section is derived by definite assignment rules for expressions.
+
+> **Open question**: We should confirm if we want to allow patterns like `var x or 1`.
```

Comment:
Ok. the correctness of such pattern would depend directly on the order of evaluation in patterns which is undefined.

So I think we should require both sides to define the same set of variables but that could be spec'ed under redeclaration rules instead of changing def assignment section.